### DISCUSSION TITLE
General delegate types

### DISCUSSION THREAD

**Discussion by JustNrik** [dotnet/csharplang#4916] [Language Ideas]
The main idea is to ease the usage of delegates without having to deal with named delegates vs Func/Action delegates. I'm not good explaining so I will show examples:

```cs
public delegate void Writer(string text);

delegate<string, void> write = Console.Write; // main idea
Action<string> action = write; // valid, Action<string> to delegate<string, void> also valid
Writer writer = write; // valid, Writer to delegate<string, void> also valid
```

This would help to avoid issues between different delegates with the same signature, like Predicate\<T> vs Func<T, bool>, custom delegates or similars.

How would this work under the hood? The idea I currently have is that it would introduce a reinterpret cast under the hood, for example, you can already do this:

```cs
public delegate void Writer(string text);

Action<string> action = Console.Write; // valid
Unsafe.As<Writer>(action).Invoke("Hello, world!"); // compiles and works as intended.
```

But I guess compiler team could figure out a better way to implement it if it exists, or maybe this is already the best way (?)

Also, making alias of this types would make things easier for named delegates (similar to c++'s typedef function pointers)

```cs
using Predicate<T> = delegate<T, bool>;
// or
using Predicate<T> = delegate bool(T);
```

Also, this would enable type inference without assuming a concrete delegate type (similar to how VB.NET's anon delegates work)

```cs
var writer = (string s) => Console.Write(s); // delegate<string, void>
```

What do you guys think?

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/issues/3068 and https://github.com/dotnet/csharplang/issues/2766#issuecomment-521645473

> ```cs
> Unsafe.As<Writer>(action).Invoke("Hello, world!"); // compiles and works as intended.
> ```

This has been suggested before. The problem is that it will work, right up until you try to call `Delegate.Combine` at which point it explodes.  The runtime still does know the type and will check it.

This kind of suggestion just isn't feasible until the runtime has something like signature equivalence for delegates.  Even then, as delegates are nominal types, you'd still need to declare the type at any API boundary.

**Reply by JustNrik** [reply]
I see... Unfortunate.

Can they make a sort of api to make this work though? for example, creating `Delegate<T>` classes inheriting Delegate and `delegate<>` would point to such classes, and they would implement helper methods to work around this and keeping the runtime type-safety? though, this would add allocations. 

Perhaps a struct that gets the target and method pointer and behaves as a sort of middleman to reduce allocations? Like, the compiler calls a sort of helper methods to pass the delegate data into the struct, the struct behaves as the delegate and could contain a lazy factory method to attempt to create the new delegate if Delegate.Combite requires it, if it fails it fails, but at least would work as a work-around, I don't know if I'm explaining myself

**Reply by HaloFour** [reply]
I'd think that if the language team wanted to solve for signature equivalency for delegates they would do so with the runtime rather than trying to figure out a way to hack it purely within the language.

Also, remember that all delegates are multicast.  Any given instance could point to more than one target.

**Comment by Rekkonnect** [comment]
Since we're already touching unsafe, I'll drop my experience with unsafe operations on delegates: I've built a library that contains such very unsafe operations and one of them is changing the runtime type of the object (remember; not stable and definitely unsafe for the most part), where you can literally change the type of the delegate and you can have it appear as a normal delegate type of your choice, and will work nicely as long as the underlying signature of the invocable delegate matches that of the actual instance's.

So for example, say you declare the following delegate:
```csharp
public delegate void ActionDoubleString(string a, string b);
```
This delegate has a signature equivalent to `Action<string, string>`, but the compiler will definitely not like it. Unless you use `Unsafe.As` or actually change the runtime type of your `ActionDoubleString` instance, you cannot operate the delegate instance the same way you would with the first-class type `Action<string, string>`, even if invoking it is the exact same process and would yield the exact same result.

And I firmly believe this is a fundamental problem with delegates; they do not boil down to a signature, but are rather fully defined types that could represent the exact same signature in different places, incompatible with one another due to lack of inheritance as well. A solution I would like is to detect the signature of the delegate type at compile time and allow compatibility with other deleagate types sharing the same signature, either with a special casting or implicitly, so as to show that delegates are actually method reference containers (multicast or not) and not raw instances that happen to contain method references.

**Reply by HaloFour** [reply]
Delegates are nominal types and there's not much that the language can do to change that, at least not without a ton of potential warts waiting to cause runtime errors.  Something like signature equivalence would have to be addressed in the runtime if it were to happen.

**Reply by TahirAhmadov** [reply]
I think it's a feature, not a bug. Signature compatibility does not make two separate delegate types equal and interchangeable.
```cs
public delegate void WriteStringToOutputDelegate(string s);
// elsewhere...
public delegate void SaveStringToFileDelegate(string s);
// ....
void Foo(WriteStringToOutputDelegate writeStringToOutput) { ... }
// ...
SaveStringToFileDelegate saveStringToFile = s => { ... };
// ...
Foo(saveStringToFile); // error - GOOD!
```
In the above example, we accidentally tried passing a delegate which isn't supposed to be used in that context, and because of type safety, were prevented from doing so. Whereas with "signature" approach, this evil bug could have gone to production.
If you specifically want signature compatibility, you can already use `Action<>` types to your heart's content. 
And if you have a 3rd party API which expects a different delegate type, you can convert them easily:
```cs
Foo(new WriteStringToOutputDelegate(saveStringToFile)); // no error - provided you know this is what you want to accomplish
```

**Reply by Rekkonnect** [reply]
If your delegate types reflect semantics about the operation that's being done, you're probably doing something wrong and what you need is a class wrapping the delegate be passed instead. Delegate types only know the signature of the methods they're invoking, nothing more and nothing less.

**Reply by HaloFour** [reply]
> If your delegate types reflect semantics about the operation that's being done, you're probably doing something wrong and what you need is a class wrapping the delegate be passed instead. Delegate types only know the signature of the methods they're invoking, nothing more and nothing less.

Delegates are nominal types, and that was intentional, so their name does hold meaning.
